extensions/amp-ad/0.1/amp-ad-custom.js

```diff
@@ -87,30 +88,94 @@ export class AmpAdCustom extends AMP.BaseElement {
           Services.xhrFor(this.win).fetchJson(fullUrl).then(res => res.json());
     }
     return ampCustomadXhrPromises[fullUrl].then(data => {
-      const element = this.element;
       // We will get here when the data has been fetched from the server
       let templateData = data;
       if (this.slot_ !== null) {
         templateData = data.hasOwnProperty(this.slot_) ? data[this.slot_] :
           null;
       }
-      // Set UI state
-      if (templateData !== null && typeof templateData == 'object') {
-        this.renderStarted();
+
+      if (!templateData || typeof templateData != 'object') {
+        this.uiHandler.applyNoContentUI();
+        return;
+      }
+
+      templateData = this.handleTemplateData_(templateData);
+
+      this.renderStarted();
+
+      try {
         Services.templatesFor(this.win)
-            .findAndRenderTemplate(element, templateData)
+            .findAndRenderTemplate(this.element, templateData)
             .then(renderedElement => {
               // Get here when the template has been rendered
-              // Clear out the template and replace it by the rendered version
-              removeChildren(element);
-              element.appendChild(renderedElement);
+              // Clear out the child template and replace it by the rendered version
+              // Note that we can't clear templates that's not ad's child because
+              // they maybe used by other ad component.
+              removeChildren(this.element);
+              this.element.appendChild(renderedElement);
+              this.signals().signal(CommonSignals.INI_LOAD);
             });
-      } else {
+      } catch (e) {
+        // Throw user error only when the element template attribute being set.
         this.uiHandler.applyNoContentUI();
+        const templateId = this.element.getAttribute('template');
+        if (templateId) {
+          this.user().error(TAG_AD_CUSTOM,
+              `Invalid templateId: ${templateId}`, e);
+        }
       }
     });
   }
 
+  /**
+   * Handles the template data response.
+   * There are two types of templateData format
+   * Format option 1
+   * {
+   *   'templateId': {},
+   *   'vars': {},
+   *   'data': {
+   *     'a': '1',
+   *     'b': '2'
+   *   }
+   * }
+   * or format option 2
+   * {
+   *  'a': '1',
+   *  'b': '2'
+   * }
+   * if `templateId` or `vars` are not specified.
+   *
+   * @param {!JsonObject} templateData
+   * @return {!JsonObject}
+   */
+  handleTemplateData_(templateData) {
+
+    let data = templateData;
+
+    if (templateData['data'] && typeof templateData['data'] == 'object') {
```

Comment:
it will be safer to check the existence of `template` attribute or `<template>` child to  differentiate the format option.

Reply:
I agree the current approach is a bit unsafe. But I assume that's super rare. 
 
One question regarding the proposed approach: In special case where `template` attribute or `<template>` child exists, and the response contains `templateId` and `data` as an object.  Do we treat the `templateData['data']` as dataObject? I would imagine that the `templateData['data']` is for template of `templateId`, not for the `<template>` child. Or should we just try to render the template with `templateData` anyway. 

I think that allowing response with `templateId` even with `<template>` child will provide more flexibility. Users get to use the template with `templateId` if the server supports the new format, and then use the child template by default w/o server support. But I am also uncertain if that will ever be the case. 

Let me know which approach to go with : )

Reply:
This "flexibility" seems too "error-prone".

The client tagging should have the overall control: if it specifies a template, it's expecting a direct data object. if it does not specifies a template, it's expecting server to tell it which template to use.